model/modeldecoder/v2/model.go

```diff
@@ -0,0 +1,463 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"net"
+
+	"github.com/minio/simdjson-go"
+	"k8s.io/apimachinery/pkg/util/json"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+
+	"github.com/elastic/apm-server/decoder"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/model/modeldecoder/validation"
+)
+
+func init() {
+	simdjsonSupported = simdjson.SupportedCPU()
+}
+
+var (
+	simdjsonSupported bool
+)
+
+//TODO(simitt): Introduce ZeroMethods to structs for reuse
+//TODO(simitt): check if required fields in Service should also be required for profiles
+type EmbeddedMetadata struct {
+	Metadata
+}
+
+type NestedMetadata struct {
+	Metadata Metadata `json:"metadata"`
+}
+
+type metadata struct {
+	Cloud   MetadataCloud   `json:"cloud"`
+	Labels  common.MapStr   `json:"labels"` //TODO(simitt): validate:"patternKeys=^[^.*\"]*$,typeVals=[string,boolean,number],maxVals=1024"`
+	Process MetadataProcess `json:"process"`
+	Service MetadataService `json:"service" validate:"required"`
+	System  MetadataSystem  `json:"system"`
+	User    MetadataUser    `json:"user"`
+}
+
+// Metadata contains event metadata
+type Metadata struct {
+	metadata
+	isSet bool
+}
+
+type metadataCloud struct {
+	Account          MetadataCloudAccount  `json:"account"`
+	AvailabilityZone validation.String     `json:"availability_zone" validate:"max=1024"`
+	Instance         MetadataCloudInstance `json:"instance"`
+	Machine          MetadataCloudMachine  `json:"machine"`
+	Project          MetadataCloudProject  `json:"project"`
+	Provider         validation.String     `json:"provider" validate:"required,max=1024"`
+	Region           validation.String     `json:"region" validate:"max=1024"`
+}
+
+// MetadataCloud holds information about the cloud computing environment
+// in which a service is running.
+type MetadataCloud struct {
+	metadataCloud
+	isSet bool
+}
+
+type metadataCloudAccount struct {
+	ID   validation.String `json:"id" validate:"max=1024"`
+	Name validation.String `json:"name" validate:"max=1024"`
+}
+
+// MetadataCloudAccount contains the name and an ID of the account
+type MetadataCloudAccount struct {
+	metadataCloudAccount
+	isSet bool
+}
```

Comment:
I am also not happy with this solution and it added some allocations, so will dig into this some more.

Reply:
The underlying types I used are `nullable` types, see [validation.String](https://github.com/elastic/apm-server/pull/4076/files/b0d1248b1f0fd5638145de995434cf557f53a95e#diff-18462a16f81a93a6983ceae8d7c3f0e9R26). I refactored and switched the logic to an `isSet()` method on the custom types to avoid the marshalling code.

Reply:
Oh right, sorry :man_facepalming: 

Given that we're using nullable types, can we get rid of `isSet()` altogether and simply compare values to the type's zero value? Is there currently a difference between, say, `metadataService.isSet()` and `metadataServiceValue != metadataService{}`?

Reply:
If the struct contains a map or array it cannot be compared with the literal; since this logic will be generated I think it'd make sense to abstract the concrete comparison in a `IsSet` or `Nil` function.

Reply:
> If the struct contains a map or array it cannot be compared with the literal

Ah of course - thanks.

> since this logic will be generated I think it'd make sense to abstract the concrete comparison in a IsSet or Nil function.

Isn't this file the hand-crafted one?

Reply:
My intention is to make the type definitions handcrafted but then generate the `Nil` (or `IsSet`) methods.